pinot-core/src/test/java/org/apache/pinot/core/operator/filter/BlockDrivenAndFilterOperatorTest.java

```diff
@@ -0,0 +1,272 @@
+    int[] dataArray = new int[limit];
+
+    Set<Integer> seenNumber = new HashSet<>();
+
+    for (int i = 0; i < limit; i++) {
+      dataArray[i] = getRandomNumber(i, NUM_ROWS, seenNumber);
+    }
+
+    BitmapBasedFilterOperator bitmapBasedFilterOperator = new BitmapBasedFilterOperator(
+        ImmutableRoaringBitmap.bitmapOf(dataArray), false, totalDocs);
+
+    executeTest(bitmapBasedFilterOperator, dataArray);
+  }
+
+  private void executeTest(BaseFilterOperator baseFilterOperator, int[] addedDocIDs) {
+    Map<String, DataSource> dataSourceMap = new HashMap<>();
+    List<ExpressionContext> expressions = new ArrayList<>();
+    for (String column : _indexSegment.getPhysicalColumnNames()) {
+      dataSourceMap.put(column, _indexSegment.getDataSource(column));
+      expressions.add(ExpressionContext.forIdentifier(column));
+    }
+    int totalDocs = _indexSegment.getSegmentMetadata().getTotalDocs();
+
+    DocIdSetOperator docIdSetOperator = new DocIdSetOperator(baseFilterOperator,
+        DocIdSetPlanNode.MAX_DOC_PER_CALL);
+    ProjectionOperator projectionOperator = new ProjectionOperator(dataSourceMap, docIdSetOperator);
+    TransformOperator transformOperator = new TransformOperator(_queryContext, projectionOperator, expressions);
+    TransformBlock transformBlock = transformOperator.nextBlock();
+
+    int limit = NUM_ROWS / 2;
+    int[] dataArray = new int[limit];
+
+    Set<Integer> seenNumber = new HashSet<>();
+
+    for (int i = 0; i < limit; i++) {
+      dataArray[i] = getRandomNumber(i, NUM_ROWS, seenNumber);
+    }
+
+    int[] finalArray;
+
+    if (addedDocIDs != null) {
+      finalArray = performIntersection(dataArray, addedDocIDs);
+    } else {
+      finalArray = dataArray;
+    }
+
+    assertTrue(finalArray != null);
+
+    BitmapBasedFilterOperator bitmapBasedFilterOperator = new BitmapBasedFilterOperator(
+        ImmutableRoaringBitmap.bitmapOf(finalArray), false, totalDocs);
+    BlockDrivenAndFilterOperator blockDrivenAndFilterOperator = new BlockDrivenAndFilterOperator(
+        bitmapBasedFilterOperator, totalDocs);
+
+    Set<Integer> resultSet = new HashSet<>();
+
+    while (transformBlock != null) {
+      ((VisitableOperator) blockDrivenAndFilterOperator).accept(transformBlock);
+
+      BlockDocIdIterator blockDocIdIterator = blockDrivenAndFilterOperator.getNextBlock()
+          .getBlockDocIdSet().iterator();
+
+      int currentValue = blockDocIdIterator.next();
+
+      while (currentValue != Constants.EOF) {
+        resultSet.add(currentValue);
+
+        currentValue = blockDocIdIterator.next();
+      }
+
+      transformBlock = transformOperator.nextBlock();
+    }
+
+    assert resultSet.size() == finalArray.length;
+
+    for (int i = 0; i < finalArray.length; i++) {
+      assertTrue(resultSet.contains(finalArray[i]), "Result does not contain " + finalArray[i]);
+    }
+  }
+
+  public static int[] performIntersection(int[] a, int[] b) {
+    return Arrays.stream(Stream.of(a)
```

Comment:
Pinot coding convention doesn't allow use of `stream` api for performance reasons.